testsuite/tests/lib-random/chi2.ml

```diff
@@ -95,6 +111,175 @@ let _ =
          (let p = Int64.to_int 17766642568158577L in
           fun () -> Random.full_int (256 * p) / p)
   end;
+
+  (* [int_in_range] *)
+
+  let min_ = -214748364 in
+  let max_ = min_ + 0x1FFF_FFFF in
+  test "Random.int_in_range, range of length 2^29 (bits 0-7)"
+       (fun () -> Random.int_in_range ~min:min_ ~max:max_ - min_);
+  test "Random.int_in_range, range of length 2^29 (bits 21-28)"
+       (fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) lsr 21);
+  let min_ = -214748364 in
+  let max_ = min_ + 0x3FFF_FFFF in
+  test "Random.int_in_range, range of length 2^30 (bits 0-7)"
+       (fun () -> Random.int_in_range ~min:min_ ~max:max_ - min_);
+  test "Random.int_in_range, range of length 2^30 (bits 22-29)"
+       (fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) lsr 22);
+  let min_int31 = -0x4000_0000 in
+  let max_int31 = 0x3FFF_FFFF in
+  test "Random.int_in_range, full int31 range (bits 0-7)"
+      (fun () -> Random.int_in_range ~min:min_int31 ~max:max_int31);
+  test "Random.int_in_range, full int31 range (bits 23-30)"
+      (fun () -> (Random.int_in_range ~min:min_int31 ~max:max_int31) lsr 23);
+  test "Random.int_in_range, range of length 256*p < 2^30 (bits 0-7)"
+      (let p = 2097169 in (* prime < 2^22 *)
+       let min_ = -214748364 in
+       let max_ = min_ + (256 * p) - 1 in
+       fun () -> (Random.int_in_range ~min:min_ ~max:max_ - min_) / p);
```

Comment:
The test subtracts `min_` before dividing, so as to translate the interval to [0, 256*p). I think that addresses the issue as well.

Using `int_in_range` with an asymmetric interval was one of my goals, to avoid falling into what could conceivably be a special case.

Reply:
Ah, right, I missed that.  In which case there should be no bias.

Reply:
But now that I think of it, this is likely (?) wrong when `256 * p > max_int`. This happens in one of the later tests, depending on `Sys.int_size`.

In tests other than those with a division by p, It does not matter if the length of the interval is larger than `max_int` because, then, the subtraction `sample - min_` wraps modulo 2^int_size, and the chi2 test only cares for the lowest 8 bits.

Reply:
I fixed this issue by using an unsigned division, i.e. by interpreting the dividend as an unsigned int. It turns out this operation is already provided by the standard library for `Int32` and `Int64`, I just had to implement it (partially, and not in the most efficient manner) for `int`:

```ocaml
(* Division of [x] by [y] where [x] is interpreted as an unsigned integer.
 * This code assumes [y >= 0]. *)
let udiv x y =
  if x >= 0 then x / y
  else
    let x' = x + min_int in
    let q = (x' / y) - (min_int / y)
    and r = (x' mod y) - (min_int mod y) in
    if r < y then q else q + 1
```

The biased chi2 test were passing nonetheless, I think, because the overflowing case is very unlikely (in these tests where `256 * p > max_int`, due to the way I chose it, `p` is such that `256 * p` is close to `max_int`, so it is very unlikely to actually draw a sample such that `sample - min_` overflows). If you’re happy with this solution, we can resolve this thread.